File: athletic_analysis/core/radar.py
"""Sprint factor profile: score five sprint-mechanics factors 0–100 for the
pentagon (radar) chart.

Scores are normalized against the same level-tiered optimal bands the form
grader uses (sprint_checks), via the shared phase bucketing, so the radar can
never disagree with the findings table: a metric graded "good" scores 100,
the minor zone maps to 60–100, major to below 60, bottoming out at zero three
tolerances beyond the band. Axes with no usable data in the clip (e.g. no
max-velocity steps in a start-only clip) stay NaN rather than pretending.
"""

from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np

from athletic_analysis.core.coaching import (_fmt_range, _fmt_value,
                                             bucket_sprint_steps, median_of,
                                             sprint_checks)
from athletic_analysis.core.metrics.sprint import SprintMetrics, StepRecord

RADAR_AXES = ("Stiffness / contact", "Front-side mechanics", "Posture / trunk",
              "Foot placement", "Rhythm")

# Axis -> metric keys that feed it (cadence handled with symmetry in Rhythm).
_AXIS_KEYS = {
    "Stiffness / contact": ("contact_ms",),
    "Front-side mechanics": ("thigh",),
    "Posture / trunk": ("trunk",),
    "Foot placement": ("overstride", "knee_strike"),
    "Rhythm": ("cadence",),
}

# Symmetry index at or beyond which the symmetry sub-score hits zero.
_SYM_ZERO = 0.15


@dataclass
class AxisScore:
    name: str
    score: float  # 0..100; NaN = no data
    detail: str
    n_steps: int = 0


@dataclass
class SprintRadar:
    axes: list[AxisScore] = field(default_factory=list)  # RADAR_AXES order
    overall: float = float("nan")
    level: str = "trained"


def band_score(value: float, lo: float, hi: float, tol: float) -> float:
    """Map a measured value to 0–100 against an optimal band, anchored to the
    grader's severity semantics: in band = 100, minor zone 100→60, major zone
    60→0 (floor at three tolerances beyond the band)."""
    if not np.isfinite(value):
        return float("nan")
    d = max(lo - value, value - hi, 0.0)
    if d == 0.0:
        return 100.0
    if tol <= 0:
        return 0.0
    if d <= tol:
        return 100.0 - 40.0 * (d / tol)
    return max(0.0, 60.0 - 60.0 * ((d - tol) / (2.0 * tol)))


def symmetry_score(steps: list[StepRecord]) -> float:
    """0–100 left/right symmetry from per-side medians of step time and
    contact time; NaN when neither quantity has a finite value per side."""
    subs = []
    for attr in ("step_time_s", "contact_time_s"):
        left = median_of([getattr(s, attr) for s in steps if s.side == "left"])
        right = median_of([getattr(s, attr) for s in steps if s.side == "right"])
        mean = 0.5 * (left + right)
        if not (np.isfinite(left) and np.isfinite(right)) or mean <= 0:
            continue
        si = abs(left - right) / mean
        subs.append(100.0 * float(np.clip(1.0 - si / _SYM_ZERO, 0.0, 1.0)))
    return float(np.mean(subs)) if subs else float("nan")


def compute_sprint_radar(kpts: np.ndarray, sprint: SprintMetrics | None,
                         velocities: dict[str, np.ndarray], fps: float,
                         level: str = "trained") -> SprintRadar | None:
    if sprint is None or not sprint.steps:
        return None
    checks_by_phase = sprint_checks(level)
    buckets = bucket_sprint_steps(kpts, sprint, velocities, fps)

    radar = SprintRadar(level=level)
    for name in RADAR_AXES:
        # (score, weight, detail) per contributing (phase, metric).
        contributions: list[tuple[float, float, str]] = []
        for phase, bucket in buckets.items():
            checks = dict(checks_by_phase.get(phase, []))
            for key in _AXIS_KEYS[name]:
                check = checks.get(key)
                if check is None:
                    continue
                value = median_of(bucket.values.get(key, []))
                score = band_score(value, check.lo, check.hi, check.tol)
                if not np.isfinite(score):
                    continue
                detail = (f"{check.metric.lower()} "
                          f"{_fmt_value(value, check.unit)} vs "
                          f"{_fmt_range(check)} ({phase})")
                contributions.append((score, len(bucket.strike_frames), detail))
        if name == "Rhythm":
            sym = symmetry_score(sprint.steps)
            if np.isfinite(sym):
                contributions.append(
                    (sym, len(sprint.steps),
                     f"L/R symmetry {sym:.0f}/100 over {len(sprint.steps)} steps"))
        if contributions:
            weights = np.array([w for _s, w, _d in contributions])
            scores = np.array([s for s, _w, _d in contributions])
            axis_score = float(np.average(scores, weights=weights))
            dominant = max(contributions, key=lambda c: c[1])[2]
            detail = "; ".join(dict.fromkeys(
                [dominant] + [d for _s, _w, d in contributions]))
            n_steps = int(max(w for _s, w, _d in contributions))
        else:
            axis_score, detail, n_steps = float("nan"), "no data in this clip", 0
        radar.axes.append(AxisScore(name=name, score=axis_score,
                                    detail=detail, n_steps=n_steps))

    finite = [a.score for a in radar.axes if np.isfinite(a.score)]
    radar.overall = float(np.mean(finite)) if finite else float("nan")
    return radar

File: ios/AthleticAnalysisCore/Sources/AthleticAnalysisCore/Radar.swift
// Sprint factor profile for the pentagon chart (port of radar.py).
// Scores are normalized against the same level-tiered bands the grader uses,
// via the shared phase bucketing, so the radar can never disagree with the
// findings table.

import Foundation

public enum RadarAxis: String, CaseIterable, Sendable {
    case stiffness = "Stiffness / contact"
    case frontSide = "Front-side mechanics"
    case posture = "Posture / trunk"
    case footPlacement = "Foot placement"
    case rhythm = "Rhythm"

    var metricKeys: [String] {
        switch self {
        case .stiffness: return ["contact_ms"]
        case .frontSide: return ["thigh"]
        case .posture: return ["trunk"]
        case .footPlacement: return ["overstride", "knee_strike"]
        case .rhythm: return ["cadence"]
        }
    }
}

public struct AxisScore: Sendable {
    public var name: String
    public var score: Double  // 0..100; NaN = no data
    public var detail: String
    public var nSteps: Int
}

public struct SprintRadar: Sendable {
    public var axes: [AxisScore] = []  // RadarAxis order
    public var overall: Double = .nan
    public var level: AthleteLevel = .trained
}

public enum Radar {
    /// Symmetry index at or beyond which the symmetry sub-score hits zero.
    static let symZero = 0.15

    /// Map a measured value to 0–100 against an optimal band: in band = 100,
    /// minor zone 100→60, major zone 60→0 (floor at 3 tolerances beyond).
    public static func bandScore(_ value: Double, lo: Double, hi: Double,
                                 tol: Double) -> Double {
        guard value.isFinite else { return .nan }
        let d = max(lo - value, value - hi, 0.0)
        if d == 0.0 { return 100.0 }
        if tol <= 0 { return 0.0 }
        if d <= tol { return 100.0 - 40.0 * (d / tol) }
        return max(0.0, 60.0 - 60.0 * ((d - tol) / (2.0 * tol)))
    }

    /// 0–100 left/right symmetry from per-side medians of step time and
    /// contact time; NaN when neither quantity has a finite value per side.
    public static func symmetryScore(_ steps: [StepRecord]) -> Double {
        var subs = [Double]()
        for attr in [\StepRecord.stepTimeS, \StepRecord.contactTimeS] {
            let left = NaNMath.nanMedian(steps.filter { $0.side == .left }
                .map { $0[keyPath: attr] })
            let right = NaNMath.nanMedian(steps.filter { $0.side == .right }
                .map { $0[keyPath: attr] })
            let mean = 0.5 * (left + right)
            guard left.isFinite, right.isFinite, mean > 0 else { continue }
            let si = abs(left - right) / mean
            subs.append(100.0 * min(1.0, max(0.0, 1.0 - si / symZero)))
        }
        return subs.isEmpty ? .nan : subs.reduce(0, +) / Double(subs.count)
    }

    public static func computeSprintRadar(_ kpts: PoseSequence,
                                          sprint: SprintMetrics?,
                                          runSpeed: [Double]?,
                                          fps: Double,
                                          level: AthleteLevel = .trained) -> SprintRadar? {
        guard let sprint, !sprint.steps.isEmpty else { return nil }
        let checksByPhase = Coaching.sprintChecks(level)
        let buckets = Coaching.bucketSprintSteps(kpts, sprint: sprint,
                                                 runSpeed: runSpeed, fps: fps)

        var radar = SprintRadar(level: level)
        for axis in RadarAxis.allCases {
            // (score, weight, detail) per contributing (phase, metric).
            var contributions: [(score: Double, weight: Double, detail: String)] = []
            // Iterate phases in a fixed order for deterministic detail strings.
            for phase in [SprintPhase.drive, .acceleration, .maxVelocity] {
                guard let bucket = buckets[phase],
                      let checks = checksByPhase[phase] else { continue }
                let checkMap = Dictionary(uniqueKeysWithValues: checks)
                for key in axis.metricKeys {
                    guard let check = checkMap[key] else { continue }
                    let value = Coaching.medianOf(bucket.values[key] ?? [])
                    let score = bandScore(value, lo: check.lo, hi: check.hi,
                                          tol: check.tol)
                    guard score.isFinite else { continue }
                    let detail = "\(check.metric.lowercased()) "
                        + "\(Coaching.fmtValue(value, unit: check.unit)) vs "
                        + "\(Coaching.fmtRange(check)) (\(phase.rawValue))"
                    contributions.append((score, Double(bucket.strikeFrames.count),
                                          detail))
                }
            }
            if axis == .rhythm {
                let sym = symmetryScore(sprint.steps)
                if sym.isFinite {
                    contributions.append(
                        (sym, Double(sprint.steps.count),
                         String(format: "L/R symmetry %.0f/100 over %d steps",
                                sym, sprint.steps.count)))
                }
            }
            let axisScore: AxisScore
            if !contributions.isEmpty {
                let totalWeight = contributions.reduce(0) { $0 + $1.weight }
                let weighted = contributions.reduce(0) { $0 + $1.score * $1.weight }
                let dominant = contributions.max { $0.weight < $1.weight }!.detail
                var seen = Set<String>()
                let details = ([dominant] + contributions.map(\.detail))
                    .filter { seen.insert($0).inserted }
                axisScore = AxisScore(name: axis.rawValue,
                                      score: weighted / totalWeight,
                                      detail: details.joined(separator: "; "),
                                      nSteps: Int(contributions.map(\.weight).max() ?? 0))
            } else {
                axisScore = AxisScore(name: axis.rawValue, score: .nan,
                                      detail: "no data in this clip", nSteps: 0)
            }
            radar.axes.append(axisScore)
        }

        let finite = radar.axes.map(\.score).filter { $0.isFinite }
        radar.overall = finite.isEmpty ? .nan
            : finite.reduce(0, +) / Double(finite.count)
        return radar
    }
}
